flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogHandleReaderWithCache.java

```diff
@@ -0,0 +1,184 @@
+                                } finally {
+                                    if (handleInputStream != null) {
+                                        IOUtils.closeQuietly(handleInputStream);
+                                    }
+                                }
+
+                                return fileStream;
+                            });
+
+            FileInputStream fin = null;
+            synchronized (cache) {
+                if (refCountedFileStream.getReferenceCounter() >= 1) {
+                    fin = new FileInputStream(refCountedFileStream.getInputFile());
+                    if (offset != 0) {
+                        LOG.debug("seek to {}", offset);
+                        fin.getChannel().position(offset);
+                    }
+                    refCountedFileStream.retain();
+                } else {
+                    // cache file already be deleted, try again
+                    return apply(handle, offset);
+                }
+            }
+
+            input =
+                    new DataInputStream(new BufferedInputStream(fin)) {
+                        @Override
+                        public void close() throws IOException {
+                            super.close();
+                            refCountedFileStream.release();
+                            if (refCountedFileStream.getReferenceCounter() == 1) {
+                                cacheCleanScheduler.schedule(
+                                        () -> cleanCacheFile(dfsPath, refCountedFileStream),
+                                        cacheIdleMillis,
+                                        TimeUnit.MILLISECONDS);
+                            }
+                        }
+                    };
+        } else {
+            input = wrapAndSeek(handle.openInputStream(), offset);
+        }
+
+        return input;
+    }
+
+    private void cleanCacheFile(Path dfsPath, RefCountedBufferingFileStream refCountedFileStream) {
+        synchronized (cache) {
+            if (refCountedFileStream.getReferenceCounter() == 1) {
+                LOG.debug("clean cached file : {}", refCountedFileStream.getInputFile().getPath());
+                cache.remove(dfsPath);
+                refCountedFileStream.release();
+            }
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        cacheCleanScheduler.shutdownNow();
+        if (!cacheCleanScheduler.awaitTermination(5, TimeUnit.SECONDS)) {
+            LOG.warn(
+                    "Unable to cleanly shutdown cache clean scheduler of "
+                            + "ChangelogHandleReaderWithCache in 5s");
+        }
+
+        Iterator<RefCountedBufferingFileStream> iterator = cache.values().iterator();
+        while (iterator.hasNext()) {
+            RefCountedBufferingFileStream cacheFile = iterator.next();
+            iterator.remove();
+            while (!cacheFile.release()) {}
```

Comment:
1. Do we need to wait here at all?
2. If yes, could it be less reource-intensive, e.g. pausing for 50ms in each iteration?

Reply:
Hi @rkhachatryan , here is to delete the cache file immediately. Generally, the reference count of all cached files should be 1 when close `ChangelogHandleReaderWithCache`, so there will be only a few loops.

Reply:
Thanks for the explanation.
If it's not waiting, isn't it more clear to delete the file explicitly like this
`Files.deleteIfExists(cacheFile.getFile().toPath());`
?

Reply:
Thanks for the explanation.
If it's not waiting, isn't it more clear to delete the file explicitly like this
`Files.deleteIfExists(cacheFile.getFile().toPath());`
?

Reply:
Thanks for the explanation.
If it's not waiting, isn't it more clear to delete the file explicitly like this
`Files.deleteIfExists(cacheFile.getFile().toPath());`
?